### PR TITLE
Add SwapId interface and update packages

### LINKED ISSUE #938: Convert swap IDs to objects
Currently we model swap IDs as byte arrays, but this doesn't provide enough structure for all situations. For example, in order to uniquely identify an atomic swap on the Ethereum blockchain, one needs to know which smart contract contains the swap. (This point is especially clear when considering that—in our implementation—native ETH swaps use a different contract to ERC20 swaps.)

This issue involves changing swap IDs to be objects, which hold a byte array, but which can be extended on a per-chain basis to incorporate more identifying information where necessary.

### PR DESCRIPTION
**Pull request by willclarktech**
Closes #938

### CODE REVIEW

File: packages/iov-bns/src/bnsconnection.spec.ts [hunk 1]
```diff
@@ -1774,21 +1785,21 @@ describe("BnsConnection", () => {
     // directly. So let's just check the last information per ID.
     const latestEventPerId = new Map<string, AtomicSwap>();
     for (const event of liveView.value()) {
-      latestEventPerId.set(toHex(event.data.id), event);
+      latestEventPerId.set(toHex(event.data.id.data), event);
```
Comment (webmaster128) [inline packages/iov-bns/src/bnsconnection.spec.ts]: Arg, here we need swap ID serialization again. Could we at least make this explicit in this test file (which is BNS only):

```ts
function serializeBnsSwapId(id: SwapId): string {
  return toHex(id.data);
}
```

File: packages/iov-ethereum/src/ethereumconnection.spec.ts [hunk 2]
```diff
@@ -85,8 +85,8 @@ async function randomAddress(): Promise<Address> {
   });
 }
 
-function matchId(id: Uint8Array): (swap: AtomicSwap) => boolean {
-  return s => Encoding.toHex(s.data.id) === Encoding.toHex(id);
+function matchId(id: SwapId): (swap: AtomicSwap) => boolean {
+  return s => Encoding.toHex(s.data.id.data) === Encoding.toHex(id.data);
```
Comment (webmaster128) [inline packages/iov-ethereum/src/ethereumconnection.spec.ts]: This should use global swap ID equality

File: packages/iov-ethereum/src/ethereumconnection.ts [hunk 3]
```diff
@@ -919,7 +936,7 @@ export class EthereumConnection implements AtomicSwapConnection {
             case SwapContractEvent.Claimed: {
               const swapId = dataArray.slice(swapIdBegin, swapIdEnd) as SwapIdBytes;
               const swapIndex = accumulator.findIndex(
-                s => Encoding.toHex(s.data.id) === Encoding.toHex(swapId),
+                s => Encoding.toHex(s.data.id.data) === Encoding.toHex(swapId),
```
Comment (webmaster128) [inline packages/iov-ethereum/src/ethereumconnection.ts]: global swap ID equality

File: packages/iov-ethereum/src/ethereumconnection.ts [hunk 4]
```diff
@@ -937,7 +954,7 @@ export class EthereumConnection implements AtomicSwapConnection {
             case SwapContractEvent.Aborted: {
               const swapId = dataArray.slice(swapIdBegin, swapIdEnd) as SwapIdBytes;
               const swapIndex = accumulator.findIndex(
-                s => Encoding.toHex(s.data.id) === Encoding.toHex(swapId),
+                s => Encoding.toHex(s.data.id.data) === Encoding.toHex(swapId),
```
Comment (webmaster128) [inline packages/iov-ethereum/src/ethereumconnection.ts]: global swap ID equality

File: packages/iov-bns/src/bnsconnection.spec.ts [hunk 5]
```diff
@@ -1774,21 +1790,21 @@ describe("BnsConnection", () => {
     // directly. So let's just check the last information per ID.
     const latestEventPerId = new Map<string, AtomicSwap>();
     for (const event of liveView.value()) {
-      latestEventPerId.set(toHex(event.data.id), event);
+      latestEventPerId.set(serializeBnsSwapId(event.data.id), event);
     }
 
     expect(latestEventPerId.size).toEqual(3);
-    expect(latestEventPerId.get(toHex(id1))).toEqual({
+    expect(latestEventPerId.get(toHex(id1.data))).toEqual({
```
Comment (webmaster128) [inline packages/iov-bns/src/bnsconnection.spec.ts]: serializeBnsSwapId

File: packages/iov-bns/src/bnsconnection.spec.ts [hunk 6]
```diff
@@ -1774,21 +1790,21 @@ describe("BnsConnection", () => {
...
       kind: SwapProcessState.Claimed,
       data: open1!.data,
       preimage: preimage1,
     });
-    expect(latestEventPerId.get(toHex(id2))).toEqual({
+    expect(latestEventPerId.get(toHex(id2.data))).toEqual({
```
Comment (webmaster128) [inline packages/iov-bns/src/bnsconnection.spec.ts]: serializeBnsSwapId

File: packages/iov-bns/src/bnsconnection.spec.ts [hunk 7]
```diff
@@ -1774,21 +1790,21 @@ describe("BnsConnection", () => {
...
       kind: SwapProcessState.Claimed,
       data: open2!.data,
       preimage: preimage2,
     });
-    expect(latestEventPerId.get(toHex(id3))).toEqual({
+    expect(latestEventPerId.get(toHex(id3.data))).toEqual({
```
Comment (webmaster128) [inline packages/iov-bns/src/bnsconnection.spec.ts]: serializeBnsSwapId

**Comment by webmaster128** [comment]
Should `SwapId.prefix` already be used as part of the ID generators in @iov/ethereum? Or should we merge this as is?

**Comment by willclarktech** [comment]
I think merge this as is, and I'll add it as part of https://github.com/iov-one/iov-core/issues/821

**Pull request merged**